Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/TapToPlace/TapToPlace.cs

```diff
@@ -0,0 +1,467 @@
+            }
+        }
+
+        protected virtual void SetRotation()
+        {
+            Vector3 direction = currentRay.Direction;
+            Vector3 surfaceNormal = currentHit.normal;
+
+            if (KeepOrientationVertical)
+            {
+                direction.y = 0;
+                surfaceNormal.y = 0;
+            }
+
+            // if the object is on a surface then change the rotation according to the normal of the hit point
+            if (DidHit && rotateAccordingToSurface)
+            {
+                GoalRotation = Quaternion.LookRotation(-surfaceNormal, Vector3.up);
+            }
+            else 
+            {
+                // If (DidHit and !rotateAccordingToSurface) or if there is no raycast hit, 
+                GoalRotation = Quaternion.LookRotation(direction, Vector3.up);
+            }
+        }
+
+        private void ToggleSpatialMeshVisibility(bool spatialMeshVisibility)
+        {
+            // If the spatial mesh is already visible on start, then we do not need to toggle the spatial mesh 
+            // if the object is in the placing state
+            if (InitialSpatialMeshDisplayOption == SpatialAwarenessMeshDisplayOptions.None || InitialSpatialMeshDisplayOption == SpatialAwarenessMeshDisplayOptions.Occlusion)
+            {
+                IReadOnlyList<IMixedRealitySpatialAwarenessMeshObserver> observers = spatialAwarenessSystemAcecess.GetDataProviders<IMixedRealitySpatialAwarenessMeshObserver>();
+
+                foreach (IMixedRealitySpatialAwarenessMeshObserver observer in observers)
+                {
+                    // If the user wants the spatial mesh visible while in the placing state
+                    if (spatialMeshVisibility)
+                    {
+                        observer.DisplayOption = SpatialAwarenessMeshDisplayOptions.Visible;
+                    }
+                    else
+                    {
+                        observer.DisplayOption = InitialSpatialMeshDisplayOption;
+                    }
+                }
+            }
+        }
+
+        #region IMixedRealityPointerHandler
+
+        /// <inheritdoc/>
+        public void OnPointerDown(MixedRealityPointerEventData eventData) { }
+
+        /// <inheritdoc/>
+        public void OnPointerDragged(MixedRealityPointerEventData eventData) { }
+
+        /// <inheritdoc/>
+        public void OnPointerUp(MixedRealityPointerEventData eventData) { }
+
+        /// <inheritdoc/>
+        public void OnPointerClicked(MixedRealityPointerEventData eventData)
+        {
+
+            // When a click is called in the same second then it is a mistake and no action needs to be taken
+            Debug.Log("Time Difference: " + (Time.time - previousFrameNumber));
+            if ((Time.time - previousFrameNumber) < 1.0f)
+            {
+                Debug.Log("Double Click has been caught, no actions triggered");
```

Comment:
Why are we catching double click and eating it?

Reply:
When you air tap or use the pinch gesture, OnClick sometimes gets called twice.  Tap to place toggles between IsBeingPlaced true/false to determine the state of the game object to place.  If OnClick is called twice in a row after one click, the game object to place will not change states.  For example, if we have a game object we want to place, we click it once and the game object enters the placing state.  If a double click occurs, the object will be in the placing state and then immediately switch back to the placed state. 

The guard is there to avoid the immediate state switching after one click.